postgres/datadog_checks/postgres/statements.py

```diff
@@ -0,0 +1,191 @@
+
+    def init_pg_stat_statements_columns(self, db):
+        """
+        Load and cache the list of the columns available under the `pg_stat_statements` table. This must be
+        queried because version is not a reliable way to determine the available columns on `pg_stat_statements`.
+        The database can be upgraded without upgrading extensions, even when the extension is included by default.
+        """
+        if self._pg_stat_statements_columns is not None and self._pg_stat_statements_query_columns is not None:
+            return
+
+        query = """
+            SELECT column_name
+            FROM information_schema.columns
+            WHERE table_schema = 'public'
+            AND table_name = 'pg_stat_statements';
+            """
+        cursor = db.cursor()
+        columns = self._execute_query(cursor, query)
+        self._pg_stat_statements_columns = frozenset(column[0] for column in columns)
+
+        # Given all of the available columns, determine the columns that will actually be used to query
+        desired_columns = (
+            list(PG_STAT_STATEMENTS_METRIC_COLUMNS.keys())
+            + list(PG_STAT_STATEMENTS_OPTIONAL_COLUMNS)
+            + list(PG_STAT_STATEMENTS_TAG_COLUMNS.keys())
+        )
+        self._pg_stat_statements_query_columns = sorted(list(set(desired_columns) & self._pg_stat_statements_columns))
+
+    def collect_per_statement_metrics(self, instance, db, instance_tags):
+        if not self.config.deep_database_monitoring:
+            return
+
+        try:
+            self.__collect_per_statement_metrics(db, instance_tags)
+        except Exception:
+            logger.exception('Unable to collect statement metrics due to an error')
+
+    def __collect_per_statement_metrics(self, instance, db, instance_tags):
+        self.init_pg_stat_statements_columns(db)
+        missing_columns = PG_STAT_STATEMENTS_REQUIRED_COLUMNS - self._pg_stat_statements_columns
+        if len(missing_columns) > 0:
+            logger.warning(
+                'Unable to collect statement metrics because required fields are unavailable: %s',
+                ', '.join(list(missing_columns)),
+            )
+            return
+
+        cursor = db.cursor(cursor_factory=psycopg2.extras.DictCursor)
+
+        rows = self._execute_query(
+            cursor,
+            STATEMENTS_QUERY.format(
+                cols=', '.join(self._pg_stat_statements_query_columns),
+                pg_stat_statements_view=self.config.pg_stat_statements_view,
+            ),
+        )
+        if not rows:
+            return
+
+        def row_keyfunc(row):
+            # old versions of pg_stat_statements don't have a query ID so fall back to the query string itself
+            queryid = row['queryid'] if 'queryid' in row else row['query']
+            return (queryid, row['datname'], row['rolname'])
+
+        rows = self._state.compute_derivative_rows(rows, PG_STAT_STATEMENTS_METRIC_COLUMNS.keys(), key=row_keyfunc)
```

Comment:
General comment on the approach. This `compute_drivative_rows` function is used to compute the diff of a counter between check runs.
The agent already provides such a feature using the `instance.monotonic_count` here. The only requirement is that metrics coming from different queries should be uniquely identified with a tag and that seems to be the case already.

So it seems like a ton of logic could be removed by simply relying on `instance.monotonic_count`.

Do I miss something? Any reason why that wouldn't work?

Reply:
We did start off using `monotonic_count` but ran into a few issues and migrated to our own logic to have greater control over the edge cases:

1. When the table is truncated or stats are reset. `monotonic_count` tracks individual metrics (columns), but it does not track rows. Therefore, when stats are reset, some columns in the row may have values greater than their predecessor. 

| check run | count | time |
|-----------|-------|------|
| A               | 100     | 550  |
| B               | 25       | 600  |

e.g. stats are reset between check run A and B because the count is lower than before. `monotonic_count` will discard `postgresql.queries.count` for this check run but it will submit `postgresql.queries.time` with an incorrect value of 50. We need to discard all metrics for B - A, which means we need to store state.
    
2. The "key" for a row doesn't always match the tags we use to submit the metric. For instance MySQL rows are _supposed_ to be unique for any combination of tags (`digest`, `schema`). But we only add the tags (`schema`, `query_signature`, `query`). The digest is computed for 1024-4096 characters of the query, but the query_signature is computed only computed for the first 200 characters of a query. This will inaccurately group together queries which share the same first 200 characters.

3. Even for a given key, rows may not be unique so metrics submitted with `monotonic_count` will see spikes between the values of duplicate keys. For instance MySQL has a bug of duplicate rows in the `events_statements_summary_by_digest` where we might see rows having the same set of tags, but vastly different values: https://bugs.mysql.com/bug.php?id=79533